### PR TITLE
111 fmt

### LINKED ISSUE #111: Design/implement formatted printing
The `Format` API is powerful, but quite difficult to use correctly. Both `Format` and `Printf` are quite complicated to implement from a type safety perspective, and format string ergonomics are less than optimal. Furthermore, the APIs make effect tracking extremely unwieldy. Design and implement a formatted printing API that cleanly supports custom formatting of inputs, effects tracking, file/stream/string output modes, ideally without excessive cognitive load on the programmer using the facility. This facility may best be implemented with explicit syntax support for format strings, in which case it won't be feasible to use in the bootstrap `Basis` library. But if no special syntax is required, favor implementing this in the bootstrap `Basis` in order to ease the OCaml -> Hemlock rewrite step of bootstrapping.

### Tasks:
- [x] Design formatter API.
- [x] Design formatted string syntax.
- [x] Extend design documentation for interpolated string syntax.
  + [x] Desugared form as semantic meaning
  + [x] `\\%` protection
- [x] Implement the `Fmt` module.
- [x] Implement `String.Fmt`.
- [x] Implement `File.Fmt`.
- [x] Implement `to_string` and `fmt` for all atomic types supported by formatted strings.
  + [x] `bool`
  + [x] `Uns`
  + [x] `Sint`
  + [x] `Intnb`-based types
  + [x] `Intw`-based types
  + [x] `real`
  + [x] `codepoint`
  + [x] `string`
- [x] Implement `fmt` for container types.
  + [x] `list`
  + [x] `array`
  + [x] `option`
  + [x] `result`
  + [x] `Ordmap.t`
  + [x] `Ordset.t`
  + [x] `Map.t`
  + [x] `Q.t`
  + [x] `Deq.t`
  + [x] `Stream.t`
- [x] Implement `fmt` for various types which provide `pp`.
  + [x] `unit`
  + [x] `Cmp.t`
  + [x] `Sign.t`
  + [x] `Real.Dir.t`
  + [x] `Real.Class.t`
  + [x] `Real.Parts.t`
  + [x] `Bytes.Slice.t`
  + [x] `Text.Cursor.t`
  + [x] `Realer.t`
- [x] Add `fmt` to interfaces in `FormattableIntf`.
- [x] Rename existing `Format`-related `pp*` symbols to `xpp*`.
- [x] Rename `fmt` in `FormattableIntf` to `pp`.
- [x] Refactor tests to use `Fmt` rather than `Format`.
  + [x] #190
  + [x] #192
  + [x] #176
  + [x] #180
  + [x] #200
  + [x] #203
  + [x] #181
  + [x] #177
  + [x] #206
  + [x] #209
  + [x] #207
  + [x] #208
  + [x] #205
  + [x] #191
  + [x] #197
  + [x] #189
  + [x] #195
  + [x] #183
  + [x] #184
  + [x] #202
  + [x] #173
  + [x] #198
  + [x] #193
  + [x] #204
  + [x] #196
  + [x] #201
  + [x] #199
  + [x] #179
  + [x] #214
  + [x] #211
  + [x] #216
  + [x] #212
  + [x] #213
  + [x] #210
  + [x] #175
  + [x] #185
  + [x] #186
  + [x] #188
- [x] Remove `xpp` from interfaces in `FormattableIntf`.
- [x] Remove all `Format`-related `xpp*` pretty printer functionality.

### PR DESCRIPTION
**Pull request by jasone**
Will resolve #111.

### CODE REVIEW

**Comment by jasone** [comment]
I made another full review pass through this and, as they say, LGTM.

File: doc/design/syntax.md [hunk 1]
```diff
@@ -405,6 +387,192 @@ context:
...
+    * `d`: Decimal
+    * `x`: Hexadecimal
+  + `<notation>` (`?notation:Fmt.notation`): Real-specific notation (default: `Fmt.Compact`)
+    * `m`: Normalized scientific form, i.e. decimal exponential or binary floating point notation
+      (mnemonic: Mantissa × base <sup>exponent</sup>)
+    * `a`: Radix point form (mnemonic: rAdix point)
+    * `c`: Trailing zeros omitted, and the radix point omitted in normalized form unless followed by
+      non-zero mantissa digits (mnemonic: Compact)
+      - Binary/octal/hexadecimal with unspecified precision: Normalized form
+      - Otherwise: The more compact of normalized vs radix point forms
```
[the remark this answers is not in the document]
Comment (jasone) [reply doc/design/syntax.md:541]: Yeah, an earlier iteration of this used some convoluted logic to decide when to use `Normalized` form with `Limited` precision for non-decimal bases to assure that by default no significant non-zero bits of the mantissa were omitted. The current iteration instead uses a straightforward orthogonal (with respect to the various formatting parameters) logic, with the caveat that the default precisions for non-decimal bases are carefully chosen such that if `RadixPoint` notation is more compact than `Normalized` notation, it still formats all of the significant non-zero mantissa bits. What we have now is more consistent and simpler to understand; the tradeoff is that if you want rounded output for non-decimal bases you have to explicitly reduce `precision`.

File: bootstrap/src/basis/intnb.ml [hunk 2]
```diff
@@ -230,9 +230,57 @@ module MakeCommon (T : ICommon) : SCommon with type t := uns = struct
...
       narrow (Int64.of_string s)
 
-    let to_string t =
+    let to_string ?(sign=Fmt.sign_default) ?(alt=Fmt.alt_default) ?(zpad=Fmt.zpad_default)
+      ?(width=Fmt.width_default) ?(base=Fmt.base_default) t =
       assert Stdlib.(Int64.(compare (narrow t) t) = 0);
-      Format.asprintf "%a" pp t
+      let rec fn accum ndigits is_neg t = begin
+        match Stdlib.(Int64.(unsigned_compare t 0L) = 0)
+              && Stdlib.(not zpad || (ndigits >= (Int64.to_int width))) with
```
[the remark this answers is not in the document]
Comment (jasone) [reply bootstrap/src/basis/intnb.ml:183]: As I've become more comfortable with how function application syntax works in OCaml I've trended toward omitting parentheses in such cases. The `ndigits >= (Int64.to_int width)` portion of that expression is an example of where parentheses are important. Similarly, if `zpad` were a more complicated expression, I would write `not (f x)`. But `(not (f x)) || ...` doesn't clarify anything about evaluation order, so I leave out the parentheses. Does that seem reasonable?

File: bootstrap/src/basis/fmt.mli [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+(** [fmt ~pad ~just ~width s formatter] creates a string based on [s] that is [~just]-justified
+    with minimum [~pad]-padded [~width] and applies the result to [formatter]. *)
+
+val sync: (module Formatter) -> (module Formatter) synced
+(** [sync formatter] calls [formatter]'s [sync] function and returns the result. A string formatter
+    returns [To_string string], whereas a file formatter returns [Synced formatter]. *)
+
+val flush: (module Formatter) -> (module Formatter)
+(** [flush formatter] calls [formatter]'s [sync] function and returns [formatter'] if [sync]
+    returns [Second formatter'], or halts if [sync] returns [First string]. *)
```
[the remark this answers is not in the document]
Comment (jasone) [reply bootstrap/src/basis/fmt.mli]: Fixed.

File: bootstrap/src/basis/fmt.mli [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+val to_string: (module Formatter) -> string
+(** [to_string formatter] calls [formatter]'s [sync] function and returns the resulting string if
+    [sync] returns [First string], or halts if [sync] returns [Second formatter']. *)
```
[the remark this answers is not in the document]
Comment (jasone) [reply bootstrap/src/basis/fmt.mli]: Fixed.

File: bootstrap/src/basis/intnb.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (cevans87) [reply bootstrap/src/basis/intnb.ml:183]: Likely reasonable. This is my old C++ battle paranoia showing up. If OCaml is sane, might as well do it the prettier way.

**Review by cevans87 — APPROVED** [review]
Nice.

File: bootstrap/test/basis/real/test_fmt.ml [hunk 5]
```diff
@@ -0,0 +1,52 @@
+open! Basis.Rudiments
+open! Basis
+open Real
+
+(* XXX Remove verbose? *)
```
[the remark this answers is not in the document]
Comment (jasone) [reply bootstrap/test/basis/real/test_fmt.ml]: This goes away later in the commit series.

File: bootstrap/test/basis/array/test_cursor.ml [hunk 6]
```diff
@@ -1,28 +1,54 @@
...
-open Format
 
 let rec fn arr hd cursor tl =
   let index = Cursor.index cursor in
-  printf "index=%a" Uns.pp index;
-  printf ", container %a arr" Cmp.pp (cmp Uns.cmp (Cursor.container cursor) arr);
+  let _ =
+    File.Fmt.stdout
+    |> Fmt.fmt "index="
+    |> Uns.fmt index
```
[the remark this answers is not in the document]
Comment (jasone) [reply bootstrap/test/basis/array/test_cursor.ml]: See #217 for design enhancements based on this idea.

**Pull request merged**